### ISSUE TITLE
why Missing ref owner context. ref cannot be used on hoisted vnodes. A vnode with ref must be created inside the render function

### ISSUE THREAD

**Issue by denisdeng** [konvajs/vue-konva#228]
<template>
    <layout-level2
        :title="title"
        :parentRoute="parentRoute"
        class="box-pdf-list"
        v-loading="loading.detail"
    >
        <template #header-right>
            <el-button type="primary" @click="finish">
                <el-icon class="mr-0.5">
                    <Collection class="text-lg" />
                </el-icon>
                完成标注
            </el-button>
        </template>

        <splitpanes class="split-pane-viewer default-theme">
            <pane class="pane-catalog" min-size="2" max-size="16">
                <ul>
                    <template v-for="(o, i) in confs" :key="i">
                        <li
                            :id="getPageId(o)"
                            :class="{ selected: pageConf.page === o.page }"
                            @click="scrollIntoPage(o)"
                        >
                            <img :src="o.thumbil" v-if="o.thumbil" />
                        </li>
                    </template>
                </ul>
            </pane>
            <pane class="pane-viewer" ref="pdfContainer">
                <div class="box-pdf-viewer" :id="id">
                    <div
                        class="pdf-viewer overflow-auto"
                        ref="container"
                        @scroll="scrollStage()"
                    >
                        <div
                            class="mx-auto overflow-hidden"
                            :style="styleObjContainer"
                        >
                            <v-stage
                                tabindex="1"
                                :ref="stage"
                                :style="offset"
                                :config="stageConf"
                                __useStrictMode
                            >
                                <v-layer :ref="layer" __useStrictMode>
                                    <template v-for="(o, i) in confs" :key="i">
                                        <v-group
                                            :config="o.group"
                                            :ref="'group-' + i"
                                        >
                                            <v-shape :config="o"></v-shape>
                                            <template
                                                v-for="(
                                                    obj, j
                                                ) in o.annotations"
                                                :key="i + '-' + j"
                                            >
                                                <v-rect
                                                    :config="obj.area"
                                                ></v-rect>
                                                <v-label :config="obj.label">
                                                    <v-tag
                                                        :config="obj.tag"
                                                    ></v-tag>
                                                    <v-text
                                                        :config="obj.text"
                                                    ></v-text>
                                                </v-label>
                                            </template>
                                        </v-group>
                                    </template>
                                    <v-transformer
                                        ref="trans"
                                        :config="transformConf"
                                    ></v-transformer>
                                </v-layer>
                            </v-stage>
                        </div>
                    </div>
                </div>
            </pane>
        </splitpanes>
    </layout-level2>
</template>
<script>
import { Rect } from 'konva'
import { Splitpanes, Pane } from 'splitpanes'
import 'splitpanes/dist/splitpanes.css'
// import * as pdfjsLib from 'pdfjs-dist/webpack'
import * as pdfjsLib from 'pdfjs-dist/build/pdf'
import * as pdfjsWorker from 'pdfjs-dist/build/pdf.worker.mjs'

pdfjsLib.GlobalWorkerOptions.workerSrc = pdfjsWorker

const labelType = {
    discard: {
        color: '#FF4D4F ',
        fill: 'rgba(255,78,79, 0.3)',
        title: '\u65E0\u6548\u533A\u57DF',
        width: 56,
        height: 24,
    },
    image: {
        color: '#FFBC17',
        fill: 'rgba(255,188,23, 0.3)',
        title: '\u56FE\u7247',
        width: 32,
        height: 24,
    },
    table: {
        color: '#52C41A',
        fill: 'rgba(82,202,20, 0.3)',
        title: '\u8868\u683C ',
        width: 32,
        height: 24,
    },
}

import { defineComponent, ref } from 'vue'
export default defineComponent({
    // eslint-disable-next-line vue/multi-word-component-names
    name: 'Annotation',
    components: {
        Splitpanes,
        Pane,
    },
    setup() {
        return {
            stage: ref('stage'),
            pdfContainer: ref('pdfContainer'),
            layer: ref('layer'),
        }
    },
    data() {
        return {
            loading: {
                detail: false,
            },
            currentObj: {
                annotation: null,
            },
            aType: 'annotation-rect',
            id: `pdf-viewer-${Date.now()}`,
            stageConf: {
                width: 1785 / 2,
                height: 984 / 2,
                listening: true,
                visible: true,
            },
            shapeConf: {
                width: 150,
                height: 150,
                canvas: null,
            },
            areaConf: {
                stroke: '#FF4D4F',
                fill: 'rgba(255,78,79, 0.3)',
                strokeWidth: 2,
            },
            textConf: {
                fontFamily: 'Calibri',
                fontSize: 14,
                padding: 5,
                fill: '#fff',
            },
            transformConf: {
                ignoreStroke: true,
                rotateEnabled: false,
            },
            scale: 1.5,
            confs: [],
            ratio: 2,
            offset: {},
            lastScrollTop: 0,
            pageConf: {
                page: 1,
                total: 1,
            },
            // stage: null,
            // layer: null,
            isDrawing: false,
            isScroll: false,
            mode: 'drawing',
            currentAnnotation: null,
            // eslint-disable-next-line @typescript-eslint/no-empty-function
            resize: () => {},
        }
    },
    computed: {
        title() {
            const {
                query: { file_name: name },
            } = this.$route
            return name
        },
        parentRoute() {
            const {
                query: { name, doc_set_id },
            } = this.$route
            return {
                name: 'DocList',
                query: {
                    name,
                    doc_set_id,
                },
                display: '',
            }
        },
        params() {
            const {
                query: { doc_id },
            } = this.$route
            return {
                doc_id: doc_id ?? '',
            }
        },
        maxPdfHeight() {
            return this.confs.reduce((a, b) => a + b.height, 0)
        },
        styleObjContainer() {
            return this.confs.length
                ? {
                      width: `${this.confs[0].width}px`,
                      height: `${this.maxPdfHeight}px`,
                  }
                : {}
        },
    },
    methods: {
        finish() {
            console.log('finish')
        },
        getPageId(o) {
            const { page } = o
            return ['page', page].join('-')
        },
        async getPage(pdf, pageNumber = 1) {
            const page = await pdf.getPage(pageNumber)
            const viewport = page.getViewport({ scale: this.scale })

            const canvas = document.createElement('canvas')
            const context = canvas.getContext('2d')

            const { width, height } = viewport
            canvas.height = height
            canvas.width = width
            const renderContext = {
                canvasContext: context,
                viewport,
            }
            await page.render(renderContext).promise
            return {
                canvas,
                width,
                height,
                page: pageNumber,
                thumbil: canvas.toDataURL(),
            }
        },
        _resize() {
            const container = document.querySelector(`#${this.id}`)
            if (container) {
                const { height, width } = container.getBoundingClientRect()
                this.stageConf.height = height
                this.stageConf.width = width
                /* if (this.stage) {
                    this.stage.height(height)
                } */
            }
        },
        getLayer() {
            const layerRef = this.$refs.layer
            return layerRef.getNode()
        },
        getStage() {
            const stageRef = this.$refs.stage
            return stageRef.getNode()
        },
        getAbsPoint(point, invert = true) {
            const layerRef = this.$refs.layer
            const layer = layerRef.getNode()
            const transform = layer.getAbsoluteTransform().copy()
            if (invert) {
                transform.invert()
            }
            return transform.point(point)
        },
        isTransforming() {
            const trans = this.getTransformer()
            return trans.isTransforming()
        },
        getTransformer() {
            const transRef = this.$refs.trans
            return transRef.getNode()
        },
        getGroupInfo(index) {
            const groupRefs = this.$refs[`group-${index}`]
            if (groupRefs && groupRefs.length) {
                const groupRef = groupRefs[0]
                return groupRef.getNode()
            }
            return null
        },
        /* selectNode () {
            const stage = this.getStage();
            let pos = stage.getPointerPosition();
            const shape = stage.getIntersection(pos);
            this.currentObj.annotation = shape;
            if (shape) {
                const width = shape.width();
                const height = shape.height();
                const x = shape.x();
                const y = shape.y();
            }
            if (!shape) {
                const trans = this.getTransformer();
                trans.detach();
            }
        },
        captureKey (evt) {
            const {keyCode} = evt;
            const {annotation: shape} = this.currentObj;
            if (keyCode === 8 && shape) {
                // const stage = this.getStage();
                const layerRef = this.$refs.layer;
                const layer = layerRef.getNode();
                const trans = this.getTransformer();
                trans.detach();
                shape.destroy();
                layer.draw();
            }
        }, */
        handleMouseDown() {
            this.currentAnnotation = null
            const stage = this.getStage()
            const layer = this.getLayer()
            this.isDrawing = true
            let pos = stage.getPointerPosition()
            const point = this.getAbsPoint(pos)
            const index = this.confs.findIndex((o, index) => {
                const {
                    attrs: {
                        data: { viewBox },
                    },
                } = this.getGroupInfo(index)
                const minY = viewBox[0].y
                const maxY = viewBox[1].y
                return point.y >= minY && point.y <= maxY
            })
            if (index !== -1) {
                const group = this.getGroupInfo(index)
                const { name } = group
                const aRects = stage.find(`.${this.aType}`)
                const id = `${name}-rect-${aRects.length}`
                const rect = new Rect({
                    ...pos,
                    width: 0,
                    height: 0,
                    fill: 'transparent',
                    stroke: labelType.discard.color,
                    strokeWidth: 2,
                    name: this.aType,
                    id,
                    // type: this.aType
                })
                group.add(rect)
                layer.batchDraw()
                this.currentAnnotation = {
                    group,
                    start: pos,
                    rect,
                }
            }
        },
        handleMouseUp() {
            this.isDrawing = false
        },
        handleMouseMove() {
            if (!this.isDrawing) {
                return
            }

            const stage = this.getStage()
            const end = stage.getPointerPosition()

            if (this.currentAnnotation) {
                const { start, rect, group } = this.currentAnnotation
                const startS = this.getAbsPoint(start)
                const endS = this.getAbsPoint(end)
                const {
                    attrs: {
                        data: { viewBox },
                    },
                } = group

                if (startS.y >= viewBox[0].y && endS.y < viewBox[1].y) {
                    const width = Math.abs(endS.x - startS.x)
                    const height = Math.abs(endS.y - startS.y)
                    rect.width(width)
                    rect.height(height)
                    rect.draggable(true)
                    rect.dragBoundFunc = pos => {
                        const absPoint = this.getAbsPoint(pos)
                        const scale = stage.getAbsoluteScale()

                        let { x, y } = absPoint
                        const offsetX = x + width * scale.x
                        const offsetY = y + height * scale.y

                        const minX = viewBox[0].x
                        const maxX = viewBox[1].x
                        const minY = viewBox[0].y
                        const maxY = viewBox[1].y

                        if (offsetX >= maxX) {
                            x -= offsetX - maxX
                        }

                        if (offsetY >= maxY) {
                            y -= offsetY - maxY
                        }

                        if (x <= minX) {
                            x = minX
                        }

                        if (y <= minY) {
                            y = minY
                        }
                        return this.getAbsPoint({ x, y }, false)
                    }
                }
            }
        },
        scrollInView(opt) {
            const obj = document.querySelector(`#${this.getPageId(opt)}`)
            if (obj) {
                obj.scrollIntoView({ behavior: 'smooth', block: 'start' })
            }
            this.pageConf.page = opt.page
        },
        scrollIntoPage({ page: index }) {
            this.isScroll = true

            const container = this.$refs.container
            const layer = this.getLayer()
            const current = this.getGroupInfo(index - 1)
            const {
                attrs: {
                    data: { viewBox },
                },
            } = current
            const yPos = viewBox[0].y * -1
            const scrollTop = yPos * -1
            container.scrollTop = scrollTop
            layer.y(yPos)
            layer.batchDraw()

            this.offset = {
                transform: `translate(0px, ${scrollTop}px)`,
            }
            this.pageConf.page = index

            setTimeout(() => {
                this.isScroll = false
            }, 300)
        },
        scrollStage() {
            if (!this.isScroll) {
                const container = this.$refs.container
                // const stageRef = this.$refs.stage;
                // const stage = stageRef.getStage();
                const stage = this.getStage()
                const layer = this.getLayer()
                const children = stage.find('.group')
                let current = null
                // const layer = this.layer;

                let { scrollTop, clientHeight } = container
                const maxOffsetY = this.maxPdfHeight * -1 + clientHeight

                let yPos = layer.y()
                const yOffset = scrollTop - this.lastScrollTop

                yPos -= yOffset
                if (yPos <= maxOffsetY) {
                    yPos = maxOffsetY
                }

                if (yPos >= 0) {
                    yPos = 0
                }

                current = children.find(o => {
                    const {
                        attrs: {
                            data: { viewBox },
                        },
                    } = o
                    const pos = yPos * -1
                    return pos > viewBox[0].y && pos < viewBox[1].y
                })

                layer.y(yPos)
                layer.batchDraw()

                if (current) {
                    const {
                        attrs: {
                            data: { page },
                        },
                    } = current
                    this.pageConf.page = page
                    setTimeout(() => {
                        this.scrollInView({ page })
                    }, 200)
                }

                this.lastScrollTop = scrollTop
                this.offset = {
                    transform: `translate(0px, ${scrollTop}px)`,
                }
            }
        },
        async init(url, roi) {
            const loadingTask = pdfjsLib.getDocument(url)
            const pdf = await loadingTask.promise
            const numPages = pdf.numPages
            this.pageConf.total = numPages
            const promises = Array.from({ length: numPages }).map(
                async (value, page) => {
                    return await this.getPage(pdf, page + 1)
                },
            )

            for (const p of promises) {
                const conf = await p
                const { page, height, width } = conf
                const x = 0
                const y = (page - 1) * height
                this.confs.push({
                    ...conf,
                    shadowColor: '#f2f4f6',
                    shadowOffset: {
                        x: 5,
                        y: 5,
                    },
                    sceneFunc: function (context) {
                        context.drawImage(conf.canvas, 0, 0)
                    },
                    annotations: roi
                        .filter(({ page_index }) => {
                            return page_index === page - 1
                        })
                        .map(({ bbox, type }) => {
                            const {
                                x: bboxX,
                                y: bboxY,
                                width: bboxW,
                                height: bboxH,
                            } = bbox
                            const calcX = width * bboxX
                            const calY = height * bboxY
                            const calcWidth = width * bboxW
                            const calcHeight = height * bboxH

                            const labelConf = labelType[type] || {}
                            return {
                                area: {
                                    ...this.areaConf,
                                    x: calcX,
                                    y: calY,
                                    width: calcWidth,
                                    height: calcHeight,
                                    stroke: labelConf.color,
                                    fill: labelConf.fill,
                                },
                                label: {
                                    x: calcX + calcWidth,
                                    y: calY - this.areaConf.strokeWidth / 2,
                                },
                                tag: {
                                    fill: labelConf.color,
                                },
                                text: {
                                    ...this.textConf,
                                    text: labelConf.title,
                                },
                            }
                        }),
                    group: {
                        name: `group`,
                        id: `group-${page}`,
                        x,
                        y,
                        data: {
                            page: page,
                            viewBox: [
                                { x, y },
                                { x: width, y: y + height },
                            ], // TODO 对应左上角和右下角坐标
                        },
                    },
                })
            }
            if (this.confs.length) {
                const { width, height } = this.confs[0]
                Object.assign(this.stageConf, { width, height })
            }

            /* this.$nextTick(() => {
                this.resize()
                const stage = this.getStage()
                const children = stage.find('.group')
                children.forEach((o, i) => {
                    setTimeout(() => {
                        this.confs[i].thumbil = o.toDataURL()
                    }, i * 10)
                })
            }) */
        },
        async getPdfInfo() {
            this.loading.detail = true
            // const pdf = await Api.Doc.getPdf(this.params)
            const { data = {} } = await Api.Doc.getDocDetail(this.params)
            const roi = _.get(data, 'area', [])
            this.loading.detail = false
            const pdfUrl = `/api/knowledge_management/doc/show?${Object.entries(
                this.params,
            )
                .map(([k, v]) => `${k}=${v}`)
                .join('&')}`
            this.init(pdfUrl, roi)
        },
    },
    created() {
        this.getPdfInfo()
    },
})
</script>
<style lang="sass">
*
    padding: 0
    margin: 0
.split-pane-viewer
    background-color: rgb(229, 231, 235)
    position: relative
    --offset-top: calc(60px + 48px + 40px)
    --height: calc(100vh - var(--offset-top))
    height: var(--height)
    overflow: hidden
    .pane-catalog
        ul, li
            list-style: none
        ul
            height: var(--height)
            overflow-y: auto
            counter-reset: order

        li
            padding: 8px 16px
            cursor: pointer
            counter-increment: order
            display: flex
            flex-direction: column
            align-items: center
            &:after
                content: counter(order)
            &.selected
                img
                    outline: 1px solid rgb(109, 98, 228)
        img
            width: 100%
            margin-bottom:  10px
            filter: drop-shadow(rgba(0, 0, 0, 0.1) 0px 1px 2px) drop-shadow(rgba(0, 0, 0, 0.06) 0px 1px 1px)
    .pane-viewer
        position: relative
.box-pdf-viewer
    // position: absolute
    height: var(--height)
    overflow: hidden
    .pdf-viewer
        margin: 0 auto
        height: var(--height)
    .overflow-auto
        overflow: auto
    .overflow-hidden
        overflow: hidden
    .mx-auto
        margin: 0 auto
</style>

**Comment by lavrton** [comment]
Can you make a very very small online demo?

**Comment by DannnyCrazy** [comment]
same error when i upgrade vue from 3.3.10 to 3.4.5

**Comment by lavrton** [comment]
https://github.com/konvajs/vue-konva/issues/229#issuecomment-1883642721
